### PR TITLE
Added support for wildcard usage in x509 SubjectAltName identification.

### PR DESCRIPTION
**Pull request by marco-vb**
Added support for single wildcard usage in x509 SubjectAltName, supports both DNS and URI.

With this change, a user can be defined as:
```xml
<user>
    <ssl_certificates>
        <subject_alt_name>URI:spiffe://foo.com/*/bar</subject_alt_name>
    </ssl_certificates>
</user>
```

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add wildcard support for user identification in x509 SubjectAltName extension.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by yakov-olkhovskiy** [comment]
This proposal requires security evaluation.
cc @santrancisco

**Comment by santrancisco** [comment]
Hi all - thank you for tagging me, @yakov-olkhovskiy . While this is more about implementation details, i did take a quick look through RFC anyway just in case and don't see any specific clause about how to implement it. 

The implementation adds more flexibility in managing endpoints communicating with Clickhouse servers when performing the same task, making it easier to manage certificates. It does add some challenge from security that we should at least make a note about it in the documentation:

 - Difficulty in Auditing - All clients will be sharing the same user to authenticate into ClickHouse database (Query log does show source IP for queries so this can be used for auditing purposes)
 - Revoking certification can be come a challenge as it would require modifying the config and revoking every other clients instead of single client
 
 But overall, i think it's a nice feature to have :)

**Comment by tonickkozlov** [comment]
@yakov-olkhovskiy @santrancisco any update please?

What can be done from our side (mine and @marco-vb's) to get this PR tested? Thank you

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit 547276780c688feed8e3dd216f661c861cdd631b with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/68236/547276780c688feed8e3dd216f661c861cdd631b/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/68236/547276780c688feed8e3dd216f661c861cdd631b/style_check.html">❌ failure</a></td></tr>
<tbody>
</table>

File: docs/en/operations/external-authenticators/ssl-x509.md [hunk 1]
```diff
@@ -6,7 +6,7 @@ import SelfManaged from '@site/docs/en/_snippets/_self_managed_only_no_roadmap.m
...
+[SSL 'strict' option](../server-configuration-parameters/settings.md#server_configuration_parameters-openssl) enables mandatory certificate validation for the incoming connections. In this case, only connections with trusted certificates can be established. Connections with untrusted certificates will be rejected. Thus, certificate validation allows to uniquely authenticate an incoming connection. `Common Name` or `subjectAltName extension` field of the certificate is used to identify the connected user. `subjectAltName extension` supports the usage of one wildcard '*' in the server configuration: see [test_x509_san_wildcard_support](../../../../tests/integration/test_ssl_cert_authentication/test.py). This allows to associate multiple certificates with the same user. Additionally, reissuing and revoking of the certificates does not affect the ClickHouse configuration.
```
Comment (yakov-olkhovskiy) [inline docs/en/operations/external-authenticators/ssl-x509.md:9]:
```suggestion
[SSL 'strict' option](../server-configuration-parameters/settings.md#server_configuration_parameters-openssl) enables mandatory certificate validation for the incoming connections. In this case, only connections with trusted certificates can be established. Connections with untrusted certificates will be rejected. Thus, certificate validation allows to uniquely authenticate an incoming connection. `Common Name` or `subjectAltName extension` field of the certificate is used to identify the connected user. `subjectAltName extension` supports the usage of one wildcard '*' in the server configuration. This allows to associate multiple certificates with the same user. Additionally, reissuing and revoking of the certificates does not affect the ClickHouse configuration.
```

**Comment by marco-vb** [comment]
hey @yakov-olkhovskiy thank you for reviewing these and testing. Does this approach look good to you overall? Is it okay to merge? Not sure about the red checks, seems like some mutation tests are not passing and not sure about the performance ones, as my change should not impact query performance.

**Comment by yakov-olkhovskiy** [comment]
@marco-vb flaky test is failing:
https://s3.amazonaws.com/clickhouse-test-reports/68236/d6ea08e8122c2577b5c486fae4e880217908ace9/integration_tests_flaky_check__asan_/integration_run_flaky_0.log
but it seems like a general flaw in the CI setup... will investigate today and probably merge as is...

**Comment by yakov-olkhovskiy** [comment]
I don't think it's relevant. Merging.

**Pull request merged**